cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/executor/DatasetAdminService.java

```diff
@@ -77,41 +80,52 @@ public boolean exists(Id.DatasetInstance datasetInstanceId) throws Exception {
    * @return dataset specification
    * @throws Exception
    */
-  public DatasetSpecification create(Id.DatasetInstance datasetInstanceId, DatasetTypeMeta typeMeta,
-                                     DatasetProperties props, boolean existing) throws Exception {
-    LOG.info("Creating dataset instance {}, type meta: {}, props: {}", datasetInstanceId, typeMeta, props);
+  public DatasetSpecification createOrUpdate(Id.DatasetInstance datasetInstanceId, DatasetTypeMeta typeMeta,
+                                             DatasetProperties props, @Nullable DatasetSpecification existing)
+    throws Exception {
+
+    if (existing == null) {
+      LOG.info("Creating dataset instance {}, type meta: {}, props: {}", datasetInstanceId, typeMeta, props);
+    } else {
+      LOG.info("Updating dataset instance {}, type meta: {}, existing: {}, props: {}",
+               datasetInstanceId, typeMeta, existing, props);
+    }
     try (DatasetClassLoaderProvider classLoaderProvider =
            new DirectoryClassLoaderProvider(cConf, locationFactory)) {
-      DatasetType type = dsFramework.getDatasetType(typeMeta, null, classLoaderProvider);
 
+      DatasetType type = dsFramework.getDatasetType(typeMeta, null, classLoaderProvider);
       if (type == null) {
         String msg = String.format("Cannot instantiate dataset type using provided type meta: %s", typeMeta);
-        LOG.error(msg);
         throw new BadRequestException(msg);
       }
+      DatasetSpecification spec = existing == null
+          ? type.configure(datasetInstanceId.getId(), props)
+          : type.reconfigure(datasetInstanceId.getId(), props, existing);
 
-      DatasetSpecification spec = type.configure(datasetInstanceId.getId(), props);
       DatasetContext context = DatasetContext.from(datasetInstanceId.getNamespaceId());
       DatasetAdmin admin = type.getAdmin(context, spec);
-      try {
+      if (existing != null && admin instanceof Updatable) {
+        ((Updatable) admin).update(existing);
+      } else {
         admin.create();
+      }
 
-        writeSystemMetadata(datasetInstanceId, spec, props, typeMeta, type, context, existing);
+      writeSystemMetadata(datasetInstanceId, spec, props, typeMeta, type, context, existing != null);
+      return spec;
 
-        return spec;
-      } catch (IOException e) {
-        String msg = String.format("Error creating dataset \"%s\": %s", datasetInstanceId, e.getMessage());
-        LOG.error(msg, e);
-        throw new IOException(msg, e);
+    } catch (Exception e) {
+      if (e instanceof IncompatibleUpdateException) {
+        // this is expected to happen if user provides bad update properties, so we log this as debug
+        LOG.debug("Incompatible update for dataset '%s'", datasetInstanceId, e);
+      } else {
+        LOG.error(String.format("Error %s dataset '%s': %s", existing == null ? "creating" : "updating",
```

Comment:
With Logger, actually we don't need to use `String.format`, we can use `{}` placeholder.